### PR TITLE
Require Julia v1.7

### LINKED ISSUE #170: `UndefVarError: RegistrySpec not defined` when trying to `run_wizard`
https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/0820b5990f2eec060f678a053a11c3e9580c54a3/src/Dependencies.jl#L164

```julia
julia> versioninfo()
Julia Version 1.7.0-beta3
Commit e76c9dad42 (2021-07-07 08:12 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin20.5.0)
  CPU: Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.0 (ORCJIT, cyclone)
Environment:
  JULIA_NUM_THREADS = 4
  JULIA_PKG_SERVER = [private registry url]
```

It looks like `Pkg.Types.RegistrySpec` is actually `Pkg.RegistrySpec` for me. 

I've tried this on the following:
```julia
julia> versioninfo()
Julia Version 1.6.1
Commit 6aaedecc44 (2021-04-23 05:59 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake-avx512)
Environment:
  JULIA_NUM_THREADS = 8
```
 and did not hit the issue, because it looks like `Pkg.Types` exports `RegistrySpec` in 1.6.1

### PR DESCRIPTION
**Pull request by giordano**
It's only one job, but in this way we can catch early failures on nightly

Fix #170, fix https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/1130, fix https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/1144

### CODE REVIEW

**Comment by giordano** [comment]
Cool, we have already failures...

**Comment by giordano** [comment]
Uhm, the test at
https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/dd83f8b0189ccda4d01dd213c6c3dd089f31065a/test/dependencies.jl#L156-L165
is failing because it installs GMP v6.2.1 (which has `libgmp.so.10.4.1`), which however [isn't compatible](https://github.com/JuliaRegistries/General/blob/28c9db36bf0f20f02c08ad00782e05469b8f839f/G/GMP_jll/Compat.toml#L10-L11) with Julia v1.6.

**Comment by giordano** [comment]
The failing test I [reported above](https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/129#issuecomment-812259489) is still failing now.  That has to be fixed in Pkg, otherwise we can't move on here.

**Comment by giordano** [comment]
Ok, after last fixes for `VersionSpec`, we're back to failures only in `test/dependencies.jl` (I can confirm all other tests run successfully):
```julia
 Downloading artifact: Zlib
 Downloading artifact: LibCURL
Setup: Test Failed at /home/runner/work/BinaryBuilderBase.jl/BinaryBuilderBase.jl/test/dependencies.jl:135
  Expression: "libssh2." * platform_dlext(platform) in readdir(last(libdirs(Prefix(destdir(dir, platform)))))
   Evaluated: "libssh2.so" in ["libcurl.la", "libcurl.so", "libcurl.so.4", "libcurl.so.4.7.0", "pkgconfig"]
Stacktrace:
 [1] macro expansion
   @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.8/Test/src/Test.jl:479 [inlined]
 [2] (::var"#37#53")(dir::String)
   @ Main ~/work/BinaryBuilderBase.jl/BinaryBuilderBase.jl/test/dependencies.jl:135
 Downloading artifact: GMP
Setup: Test Failed at /home/runner/work/BinaryBuilderBase.jl/BinaryBuilderBase.jl/test/dependencies.jl:176
  Expression: isfile(joinpath(destdir(dir, platform), "lib", "libgmp.so.10.4.0"))
Stacktrace:
 [1] macro expansion
   @ /opt/hostedtoolcache/julia/nightly/x64/share/julia/stdlib/v1.8/Test/src/Test.jl:479 [inlined]
 [2] (::var"#45#61")(dir::String)
   @ Main ~/work/BinaryBuilderBase.jl/BinaryBuilderBase.jl/test/dependencies.jl:176
Test Summary:                     | Pass  Fail  Broken  Total   Time
Dependencies                      |   72     2       2     76  52.5s
  Filter dependencies by platform |    2                    2   0.2s
  JSON (de)serialization          |   10             1     11   1.8s
  Setup                           |   22     2       1     25  49.6s
```

However I'm confused by why we're getting the wrong GMP in nightly, with a recent built from `master` I get:
```julia
julia> Pkg.add("GMP_jll"; julia_version=v"1.6")
   Resolving package versions...
    Updating `/tmp/jl_fGToRx/Project.toml`
  [781609d7] ↓ GMP_jll v6.2.1+1 ⇒ v6.2.0+5
    Updating `/tmp/jl_fGToRx/Manifest.toml`
  [781609d7] ↓ GMP_jll v6.2.1+1 ⇒ v6.2.0+5

julia> Pkg.status()
Status `/tmp/jl_fGToRx/Project.toml`
  [781609d7] GMP_jll v6.2.0+5
```
while in v1.7:
```julia
julia> Pkg.add("GMP_jll"; julia_version=v"1.6")
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
    Updating `/tmp/jl_xc2Y49/Project.toml`
  [781609d7] + GMP_jll
    Updating `/tmp/jl_xc2Y49/Manifest.toml`
  [56f22d72] + Artifacts
  [8f399da3] + Libdl
  [781609d7] + GMP_jll
Precompiling project...
  1 dependency successfully precompiled in 1 seconds

julia> Pkg.status()
      Status `/tmp/jl_xc2Y49/Project.toml`
  [781609d7] GMP_jll
```
which seems to confirm that this will never work in v1.7.

We need to investigate why the build environment is getting wrong GMP_jll, and why `LibCURL_jll` dependencies doesn't seem to be automatically pulled in.

**Comment by giordano** [comment]
I tried to make package installation more verbose:
```
   Resolving package versions...
    Updating `/tmp/jl_APzAHS/x86_64-linux-gnu-julia_version+1.6.0/.project/Project.toml`
  [781609d7] + GMP_jll v6.2.0+5
    Updating `/tmp/jl_APzAHS/x86_64-linux-gnu-julia_version+1.6.0/.project/Manifest.toml`
  [56f22d72] + Artifacts
  [8f399da3] + Libdl
  [781609d7] + GMP_jll v6.2.0+5
readdir(joinpath(destdir(dir, platform), "lib")) = ["libgmp.la", "libgmp.so", "libgmp.so.10", "libgmp.so.10.4.1", "libgmpxx.la", "libgmpxx.so", "libgmpxx.so.4", "libgmpxx.so.4.6.1", "pkgconfig"]
Setup: Test Failed at /home/mose/.julia/dev/BinaryBuilderBase/test/dependencies.jl:177
  Expression: isfile(joinpath(destdir(dir, platform), "lib", "libgmp.so.10.4.0"))
Stacktrace:
 [1] macro expansion
   @ ~/repo/julia/usr/share/julia/stdlib/v1.8/Test/src/Test.jl:479 [inlined]
 [2] (::var"#17#33")(dir::String)
   @ Main ~/.julia/dev/BinaryBuilderBase/test/dependencies.jl:177
Test Summary:                     | Pass  Fail  Broken  Total   Time
Dependencies                      |   72     2       2     76  12.9s
  Filter dependencies by platform |    2                    2   0.2s
  JSON (de)serialization          |   10             1     11   1.7s
  Setup                           |   22     2       1     25  10.2s
ERROR: LoadError: Some tests did not pass: 72 passed, 2 failed, 0 errored, 2 broken.
in expression starting at /home/mose/.julia/dev/BinaryBuilderBase/test/dependencies.jl:18
in expression starting at /home/mose/.julia/dev/BinaryBuilderBase/test/runtests.jl:4
```
I really don't understand this: `Pkg.jl` says it's installing `GMP_jll` `v6.2.0+5`, which is correct (it's what's [used in Julia v1.6](https://github.com/JuliaLang/julia/blob/release-1.6/stdlib/GMP_jll/Project.toml#L3)), but  `libgmp.so.10.4.1` is in GMP 6.2.1, not 6.2.0, so this doesn't make much sense to me.

**Comment by giordano** [comment]
This
https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/030c6e47d0e2bb9f1b146b9e5a8d3231e9c437b5/src/Prefix.jl#L479-L483
is probably what is different between v1.6 and nightly.  When adding `Pkg.add("GMP_jll"; julia_version=v"1.6")`, `dep_path` is `"/home/mose/Documents/julia-1.6.4/share/julia/stdlib/v1.6/GMP_jll"` in v1.6 and `"/home/mose/repo/julia/usr/share/julia/stdlib/v1.8/GMP_jll"` is nightly.  I guess this has happened to work just by chance so far because we were on v1.6.  Not sure how this is supposed to work.  Pinging @staticfloat, you added those lines in #97

**Comment by staticfloat** [comment]
Hmmm, could `dep.tree_hash` be non-`nothing` somehow?

**Comment by giordano** [comment]
It's non-`nothing` for non-stdlibs.  `GMP_jll` happens to be an stdlib, but we want to get a version which may not be a current stdlib

**Comment by giordano** [comment]
We do something like this:
```
(@v1.8) pkg> activate --temp
  Activating new project at `/tmp/jl_0tzUjS`

julia> using Pkg

julia> ctx = Pkg.Types.Context(;julia_version=v"1.6");

julia> Pkg.add(ctx, [PackageSpec("GMP_jll"), PackageSpec("GTK3_jll")])
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
    Updating `/tmp/jl_0tzUjS/Project.toml`
  [77ec8976] + GTK3_jll v3.24.31+0
  [781609d7] + GMP_jll v6.2.0+5
    Updating `/tmp/jl_0tzUjS/Manifest.toml`
  [692b3bcd] + JLLWrappers v1.3.0
  [21216c6a] + Preferences v1.2.3
  [7b86fcea] + ATK_jll v2.36.1+0
  [6e34b625] + Bzip2_jll v1.0.8+0
  [83423d85] + Cairo_jll v1.16.1+1
  [ee1fde0b] + Dbus_jll v1.12.16+3
  [2e619515] + Expat_jll v2.2.10+0
  [a3f928ae] + Fontconfig_jll v2.13.93+0
  [d7e528f0] + FreeType2_jll v2.10.4+0
  [559328eb] + FriBidi_jll v1.0.10+0
  [77ec8976] + GTK3_jll v3.24.31+0
  [78b55507] + Gettext_jll v0.21.0+0
  [7746bdde] + Glib_jll v2.68.3+2
  [3b182d85] + Graphite2_jll v1.3.14+0
  [2e76f6c2] + HarfBuzz_jll v2.8.1+1
  [aacddb02] + JpegTurbo_jll v2.1.0+0
  [dd4b983a] + LZO_jll v2.10.1+0
  [42c93a91] + Libepoxy_jll v1.5.8+1
  [e9f186c6] + Libffi_jll v3.2.2+1
  [d4300ac3] + Libgcrypt_jll v1.8.7+0
  [7e76a0d4] + Libglvnd_jll v1.3.0+3
  [7add5ba3] + Libgpg_error_jll v1.42.0+0
  [94ce4f54] + Libiconv_jll v1.16.1+1
  [4b2f31a3] + Libmount_jll v2.35.0+0
  [89763e89] + Libtiff_jll v4.3.0+0
  [38a345b3] + Libuuid_jll v2.36.0+0
  [2f80f16e] + PCRE_jll v8.44.0+0
  [36c8627f] + Pango_jll v1.47.0+0
  [30392449] + Pixman_jll v0.40.1+0
  [a2964d1f] + Wayland_jll v1.19.0+0
  [2381bf8a] + Wayland_protocols_jll v1.23.0+0
  [02c8fc9c] + XML2_jll v2.9.12+0
  [aed1982a] + XSLT_jll v1.1.34+0
  [4f6342f7] + Xorg_libX11_jll v1.6.9+4
  [0c0b7dd1] + Xorg_libXau_jll v1.0.9+4
  [3c9796d7] + Xorg_libXcomposite_jll v0.4.5+4
  [935fb764] + Xorg_libXcursor_jll v1.2.0+4
  [0aeada51] + Xorg_libXdamage_jll v1.1.5+4
  [a3789734] + Xorg_libXdmcp_jll v1.1.3+4
  [1082639a] + Xorg_libXext_jll v1.3.4+4
  [d091e8ba] + Xorg_libXfixes_jll v5.0.3+4
  [a51aa0fd] + Xorg_libXi_jll v1.7.10+4
  [d1454406] + Xorg_libXinerama_jll v1.1.4+4
  [ec84b674] + Xorg_libXrandr_jll v1.5.2+4
  [ea2f1a96] + Xorg_libXrender_jll v0.9.10+4
  [b6f176f1] + Xorg_libXtst_jll v1.2.3+4
  [14d82f49] + Xorg_libpthread_stubs_jll v0.1.0+3
  [c7cfdc94] + Xorg_libxcb_jll v1.13.0+3
  [cc61e674] + Xorg_libxkbfile_jll v1.1.0+4
  [35661453] + Xorg_xkbcomp_jll v1.4.2+4
  [33bec58e] + Xorg_xkeyboard_config_jll v2.27.0+4
  [c5fb5394] + Xorg_xtrans_jll v1.4.0+3
  [3161d3a3] + Zstd_jll v1.5.0+0
  [de012916] + at_spi2_atk_jll v2.34.1+4
  [0fc3237b] + at_spi2_core_jll v2.34.0+4
  [da03df04] + gdk_pixbuf_jll v2.42.6+1
  [bf975903] + iso_codes_jll v4.3.0+4
  [b53b4c65] + libpng_jll v1.6.38+0
  [d8fb68d0] + xkbcommon_jll v0.9.1+5
  [0dad84c5] + ArgTools v1.1.1
  [56f22d72] + Artifacts
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [f43a241f] + Downloads v1.4.0
  [7b1f6079] + FileWatching
  [b77e0a4c] + InteractiveUtils
  [b27032c2] + LibCURL v0.6.2
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [ca575930] + NetworkOptions v1.2.0
  [44cfe95a] + Pkg v1.5.0
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [fa267f1f] + TOML v1.0.0
  [a4e569a6] + Tar v1.9.0
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [e66e0078] + CompilerSupportLibraries_jll v0.3.6+1
  [781609d7] + GMP_jll v6.2.0+5
  [deac9b47] + LibCURL_jll v7.73.0+3
  [29816b5a] + LibSSH2_jll v1.9.1+0
  [c8ffd9c3] + MbedTLS_jll v2.24.0+1
  [14a3606d] + MozillaCACerts_jll v2020.7.22
  [83775a58] + Zlib_jll v1.2.12+0
  [8e850ede] + nghttp2_jll v1.41.0+0
  [3f19e933] + p7zip_jll v16.2.1+0

julia> ctx.env.manifest
Pkg.Types.Manifest(v"1.6.0", v"2.0.0", Dict{Base.UUID, Pkg.Types.PackageEntry}(UUID("2381bf8a-dfd0-557d-9999-79630e7b1b91") => PackageEntry(
  name = Wayland_protocols_jll
  version = 1.23.0+0
  tree_hash = 66d72dc6fcc86352f01676e8f0f698562e60510f
), UUID("3c9796d7-64a0-5134-86ad-79f8eb684845") => PackageEntry(
  name = Xorg_libXcomposite_jll
  version = 0.4.5+4
  tree_hash = 7c688ca9c957837539bbe1c53629bb871025e423
), UUID("da03df04-f53b-5353-a52f-6a8b0620ced0") => PackageEntry(
  name = gdk_pixbuf_jll
  version = 2.42.6+1
  tree_hash = c23323cd30d60941f8c68419a70905d9bdd92808
), UUID("ee1fde0b-3d02-5ea6-8484-8dfef6360eab") => PackageEntry(
  name = Dbus_jll
  version = 1.12.16+3
  tree_hash = 97f1325c10bd02b1cc1882e9c2bf6407ba630ace
), UUID("aacddb02-875f-59d6-b918-886e6ef4fbf8") => PackageEntry(
  name = JpegTurbo_jll
  version = 2.1.0+0
  tree_hash = d735490ac75c5cb9f1b00d8b5509c11984dc6943
), UUID("bf975903-5238-5d20-8243-bc370bc1e7e5") => PackageEntry(
  name = iso_codes_jll
  version = 4.3.0+4
  tree_hash = 5ee24c3ae30e006117ec2da5ea50f2ce457c019a
), UUID("76f85450-5226-5b5a-8eaa-529ad045b433") => PackageEntry(
  name = LibGit2
), UUID("2a0f44e3-6c83-55bd-87e4-b1978d98bd5f") => PackageEntry(
  name = Base64
), UUID("3b182d85-2403-5c21-9c21-1e1f0cc25472") => PackageEntry(
  name = Graphite2_jll
  version = 1.3.14+0
  tree_hash = 344bf40dcab1073aca04aa0df4fb092f920e4011
), UUID("30392449-352a-5448-841d-b1acce4e97dc") => PackageEntry(
  name = Pixman_jll
  version = 0.40.1+0
  tree_hash = b4f5d02549a10e20780a24fce72bea96b6329e29
)…), Dict{String, Any}("project_hash" => "ebca7a22387a91578e1a2b38adee39a95df75fc7"))
```
How can we get the path of `GMP_jll` `v6.2.0+5`?  We don't have its git tree sha1, so we can't use `Pkg.Operations.find_installed`

**Comment by giordano** [comment]
If I use look up the git tree hash in the registry for version [`v6.2.0+5`](https://github.com/JuliaRegistries/General/blob/f06898b8c4ca48da9e447f18c8720c357acbf515/G/GMP_jll/Versions.toml#L51) I get this path
```
julia> Pkg.Operations.find_installed("GMP_jll", UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), Base.SHA1("40388878122d491a2e55b0e730196098595d8a90"))
"/home/mose/.julia/packages/GMP_jll/IA6Pi"
```
However this path doesn't exist on disk:
```
shell> ls $(Pkg.Operations.find_installed("GMP_jll", UUID("781609d7-10c4-51f6-84f2-b8444358ff6d"), Base.SHA1("40388878122d491a2e55b0e730196098595d8a90")))
ls: cannot access '/home/mose/.julia/packages/GMP_jll/IA6Pi': No such file or directory
```
So `Pkg` pretended to install `v6.2.0+5`, but it did nothing.

**Comment by staticfloat** [comment]
So the problem is that GMP_jll is a standard library on v1.6, so when you tell Pkg to `add()` something for a specific Julia version, it will do so, unless it's a standard library.  I think strictly speaking, this is desired behavior from the point of view of `Pkg`, since we can't load those other stdlib versions, and if you were on Julia v1.6, it wouldn't install a GMP_jll anyway.

So what we need to do is make use of the `julia_version = nothing` ability that I added, and just give the `PackageSpec` the proper version number:

```julia
julia> using Pkg, TOML

       function collect_jlls(manifest::Dict, dependencies::Set{<:AbstractString})
           # Handle Manifest format v2.0
           if get(manifest, "manifest_format", "1.0") == "2.0" && haskey(manifest, "deps")
               return collect_jlls(manifest["deps"], dependencies)
           end

           jlls = copy(dependencies)    
           for (pkg, infos) in manifest
               for info in infos
                   if pkg in jlls
                       deps = get(info, "deps", nothing)
                       deps === nothing && continue
                       for dep in deps
                           if endswith(dep, "_jll")
                               push!(jlls, dep)
                           end
                       end
                   end
               end
           end
           if jlls == dependencies
               return jlls
           else
               return collect_jlls(manifest, jlls)
           end
       end

       julia_version = v"1.6"
       ctx = Pkg.Types.Context(;julia_version)

       # Install something that has a bunch of dependencies, AND something that's an stdlib
       dependencies = PackageSpec.(["GMP_jll", "GTK3_jll"])
       Pkg.add(ctx, dependencies)

       # Get all JLL packages within the current project
       installed_jlls = [
           Pkg.Types.PackageSpec(name=p.name, uuid=u, tree_hash=p.tree_hash) for (u, p) in ctx.env.manifest if endswith(p.name, "_jll")
       ]

       manifest = TOML.parsefile(ctx.env.manifest_file)
       dependencies_names = Set(getfield.(dependencies, :name))
       jll_direct_deps = collect_jlls(manifest, dependencies_names)
       filter!(jll -> jll.name ∈ jll_direct_deps, installed_jlls)

       for (dep_idx, dep) in enumerate(installed_jlls)
           # First, figure out if this JLL was treated as a stdlib:
           if dep.tree_hash === nothing
               # If it was, we need to figure out the version for this julia version
               dep_version = Pkg.Types.stdlib_version(dep.uuid, julia_version)
               @show dep.name, dep_version
               
               # Then call Pkg.add() with that precise version, and julia_version == nothing
               dep.version = dep_version
               Pkg.add(ctx, [dep]; julia_version = nothing)
               installed_jlls[dep_idx] = dep
           end
       end

       # Now that we've concretized all stdlib JLLs, everybody has a path:
       for dep in installed_jlls
           dep_path = Pkg.Operations.find_installed(dep.name, dep.uuid, dep.tree_hash)
           @show dep.name, dep_path
       end
```

**Comment by staticfloat** [comment]
Highlights of the above:

1) We're calling `Pkg.add()` multiple times here; I think we can do it only twice pretty easily; once to download everything for the top-level deps, then we scan for stdlibs, fix them to their concrete version, then re-add to download them and their artifacts.  Note that I'm overriding `julia_version` to `nothing` in the second call; we want to continue to keep `julia_version` as the actual number in the first call, since that will correctly propagate julia version constrains upon all the other packages.  We ignore these for stdlibs because, well, we know exactly what version things should be for those.

2) I had to fix `collect_jlls()` because of the new manifest format.  It's as easy as recursing into the `deps` sub-key, which used to be the single top-level entity.

3) I'm kind of relying upon the fact that the second `Pkg.add()` modifies the `dep` object here, giving it the treehash that it needs for the final `find_installed()` call.  I don't think that's a good design; let's future-proof ourselves against `Pkg` changes and re-parse the Manifest after the second `Pkg.add()`?

**Comment by staticfloat** [comment]
Okay, I spent a long time figuring out how to get _exactly_ the version we want; (e.g. `1.2.12+1` and not `1.2.12+2` for Zlib_jll) but now that I'm thinking about it.... perhaps that's not so important; perhaps it's okay that the resolver will ignore the build number, because we really shouldn't be that worried about the build number, even with stdlibs..... right?  What do you guys think?

**Comment by DilumAluthge** [comment]
For non-stdlibs, if you add by the commit, e.g. `Pkg.add(rev = "my commit hash")`, then you force Pkg to give you that exact commit, which will allow you to get an older build number.

I don't think that will work for stdlibs.

**Comment by giordano** [comment]
[`get_addable_spec`](https://github.com/JuliaPackaging/Yggdrasil/blob/185b2337e1843f22529d86bd4e503a1179fc66e9/fancy_toys.jl#L49-L114) works fine with stdlibs, because it queries the registry, which is what I suggested above.  We may eventually incorporate it here

**Comment by giordano** [comment]
For the record, tests pass for me locally in Julia v1.7, apart from https://github.com/JuliaLang/Pkg.jl/issues/2931 which makes installation of a package crash, but _after_ it has been downloaded, so that subsequent calls to that function work fine: if I run the tests twice, they are eventually successful.

File: src/compat.jl [hunk 1]
```diff
@@ -16,3 +16,16 @@ function valid_dl_path(path, platform)
         return false
     end
 end
+
+# If we don't have `stdlib_version` from Pkg, recreate it ourselves
+if !isdefined(Pkg.Types, :stdlib_version)
+    function stdlib_version(uuid::Base.UUID, julia_version::Union{VersionNumber,Nothing})::Union{VersionNumber,Nothing}
+        last_stdlibs = Pkg.Types.get_last_stdlibs(julia_version)
```
Comment (giordano) [inline src/compat.jl:23]: In Julia v1.6 `Pkg.Types.get_last_stdlibs` doesn't return the version numbers, only the names.  Unless there are other ways to get the version number, I believe we're going to drop support for this version

File: test/compat.jl [hunk 2]
```diff
@@ -19,6 +21,22 @@ using Downloads: RequestError
     @test basename(strip_path_separator("/home/wizard/")) == "wizard"
     @test basename(strip_path_separator("/home//wizard///")) == "wizard"
     @test basename(strip_path_separator("wizard.jl")) == "wizard.jl"
+
+    gmp_jll = UUID("781609d7-10c4-51f6-84f2-b8444358ff6d")
+    llvmlibwnwind_jll = UUID("47c5dbc3-30ba-59ef-96a6-123e260183d9")
+    @test stdlib_version(gmp_jll, v"1.6") == v"6.2.0+5"
+    @test stdlib_version(llvmlibwnwind_jll, v"1.6") === nothing
+    if VERSION ≥ v"1.7-DEV"
+        # Fascinating, different versions of Julia have different opinions about GMP_jll in v1.7
```
Comment (giordano) [inline test/compat.jl:30]: :upside_down_face:

**Comment by giordano** [comment]
TESTS ARE PASSING IN v1.7 AND NIGHTLY!

**Comment by giordano** [comment]
Let's do it! Thanks a lot for the help!

**Pull request merged**